### PR TITLE
Centered on asteroid

### PR DESCRIPTION
**Pull request by RobinJehn**

### CODE REVIEW

**Comment by RobinJehn** [comment]
@Tait-A Please do a quick review and merge. I wouldn't worry to much about the state of train.py as mark got a rework of that coming on another branch anyhow.

File: src/centered-on-asteroid/model_def.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        """
+        Args:
+            output_size (int): The number of outputs
+            image_shape (Tuple[int, int]): Shape of the input image
+        """
+        super(CNN, self).__init__()
+        n_channels_1 = 16
+        n_channels_2 = 32
+        final_layer_size = (
+            n_channels_2 * (image_shape[0] // 2 // 2) * (image_shape[1] // 2 // 2)
```
Comment (Tait-A) [inline src/centered-on-asteroid/model_def.py]: this can be // 4

File: src/centered-on-asteroid/visualize.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+# Load model
+images_per_sequence = 4
+feature_vector_size = 10
+image_shape = (30, 30)
+
+model = model_def.CFN(images_per_sequence, feature_vector_size, image_shape)
+model.load_state_dict(torch.load(model_path))
+model.eval()
+
+df = get_dataframe("/Users/robinjehn/Documents/repos/StarShooter/data/csv/")
```
Comment (Tait-A) [inline src/centered-on-asteroid/visualize.py]: Extract paths to config for generalisation

**Review by Tait-A — DISMISSED** [review]
2 comments, in model_def.py and visualise.py

File: src/centered-on-asteroid/model_def.py (hunk 1 above)
Comment (RobinJehn) [reply src/centered-on-asteroid/model_def.py]: Wasn't super sure, because of rounding. But yeah I think you are right that // 2 // 2 = // 4, but its not always true that // x // y = // xy

**Comment by RobinJehn** [comment]
Should have addressed the comments, have a quick look over my comments and then merge.

File: src/centered-on-asteroid/train.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    run_name = "{}_{}".format(experiment_name, timestamp)
+    logger = tensorboard.writer.SummaryWriter("logs/" + run_name)
+
+    model_folder = "model/{}".format(run_name)
+    os.makedirs(model_folder)
+
+    # Needed to compute validation accuracy
+    val_images, val_labels = val_dataset[:]
+
+    best_val_loss = 1_000_000.0
```
Comment (mark-a-austin) [inline src/centered-on-asteroid/train.py]: should be torch.inf

File: src/centered-on-asteroid/train.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+    # Make sure at least one window is used. Used for inter-epoch reporting
+    window_size = min(window_size, len(training_loader))
+    for i, data in enumerate(training_loader):
+        inputs, labels = data
+        optimizer.zero_grad()  # reset
+        preds = model(inputs)
+
+        # Compute the loss and its gradients
+        loss = criterion(preds, labels.float())
```
Comment (mark-a-austin) [inline src/centered-on-asteroid/train.py:33]: Should labels not already be floats? Or add it to the preprocessing of the loader at least

File: src/centered-on-asteroid/utils.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        data_set (TensorDataset): Torch Dataset consisting of (input, output) pairs
+        split (Tuple[float, float]): Percentage used for training and validation. Should sum to 1.
+        batch_size (int): Batch size used for the training loader.
+
+    Returns: Training data loader and validation dataset
+    """
+
+    train_data_set, val_data_set = torch.utils.data.random_split(data_set, split)
+    train_loader = torch.utils.data.DataLoader(
+        train_data_set, batch_size=batch_size, shuffle=True
```
Comment (mark-a-austin) [inline src/centered-on-asteroid/utils.py:113]: perhaps set the drop_last flag on to ensure batch size consistency

File: src/centered-on-asteroid/visualize.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
+from utils import get_dataset, get_dataframe
+import torch
+import model_def
+import matplotlib.pyplot as plt
+from config import CSV_FOLDER, COA_IMAGE_FOLDER
+
+model_path = "/Users/robinjehn/Documents/repos/StarShooter/src/centered-on-asteroid/model/end-to-end_20240216_161358/model_20240216_161358_4"
```
Comment (mark-a-austin) [inline src/centered-on-asteroid/visualize.py:7]: Model path should come from config folder

**Review by mark-a-austin — CHANGES_REQUESTED** [review]

**Review by Tait-A — DISMISSED** [review]

**Review by mark-a-austin — APPROVED** [review]
